src/test/java/org/jenkins/plugins/lockableresources/LockStepTest.java

```diff
@@ -0,0 +1,216 @@
+						"	lock('resource1') {\n" +
+						"		semaphore 'wait-b'\n" +
+						"	}\n" +
+						"}\n"
+				));
+
+				WorkflowRun b1 = p.scheduleBuild2(0).waitForStart();
+				SemaphoreStep.waitForStart("wait-b/1", b1);
+				// both messages are in the log because branch b acquired the lock and branch a is waiting to lock
+				story.j.waitForMessage("[b] Lock acquired on [resource1]", b1);
+				story.j.waitForMessage("[a] [resource1] is locked, waiting...", b1);
+
+				SemaphoreStep.success("wait-b/1", null);
+
+				story.j.waitForMessage("[a] Lock acquired on [resource1]", b1);
+			}
+		});
+	}
+
+	@Test
+	public void lockOrderRestart() {
+		story.addStep(new Statement() {
+			@Override
+			public void evaluate() throws Throwable {
+				defineResource("resource1");
+				WorkflowJob p = story.j.jenkins.createProject(WorkflowJob.class, "p");
+				p.setDefinition(new CpsFlowDefinition(
+						"lock('resource1') {\n" +
+						"	semaphore 'wait-inside'\n" +
+						"}\n" +
+						"echo 'Finish'"
+				));
+				WorkflowRun b1 = p.scheduleBuild2(0).waitForStart();
+				SemaphoreStep.waitForStart("wait-inside/1", b1);
+				WorkflowRun b2 = p.scheduleBuild2(0).waitForStart();
+				// Ensure that b2 reaches the lock before b3
+				story.j.waitForMessage("[resource1] is locked, waiting...", b2);
+				WorkflowRun b3 = p.scheduleBuild2(0).waitForStart();
+				// Both 2 and 3 are waiting for locking resource1
+
+				story.j.waitForMessage("[resource1] is locked, waiting...", b3);
+			}
+		});
+
+		story.addStep(new Statement() {
+			@Override
+			public void evaluate() throws Throwable {
+				WorkflowJob p = story.j.jenkins.getItemByFullName("p", WorkflowJob.class);
+				WorkflowRun b1 = p.getBuildByNumber(1);
+				WorkflowRun b2 = p.getBuildByNumber(2);
+				WorkflowRun b3 = p.getBuildByNumber(3);
+
+				// Unlock resource1
+				SemaphoreStep.success("wait-inside/1", null);
+				story.j.waitForMessage("Lock released on resouce [resource1]", b1);
+
+				story.j.waitForMessage("Lock acquired on [resource1]", b2);
+				story.j.assertLogContains("[resource1] is locked, waiting...", b3);
+				SemaphoreStep.success("wait-inside/2", null);
+				SemaphoreStep.waitForStart("wait-inside/3", b3);
+				story.j.assertLogContains("Lock acquired on [resource1]", b3);
+				SemaphoreStep.success("wait-inside/3", null);
+				story.j.waitForMessage("Finish", b3);
+			}
+		});
+	}
+
+	@Test
+	public void interoperability() {
+		story.addStep(new Statement() {
+			@Override
+			public void evaluate() throws Throwable {
+				defineResource("resource1");
+				WorkflowJob p = story.j.jenkins.createProject(WorkflowJob.class, "p");
+				p.setDefinition(new CpsFlowDefinition(
+						"lock('resource1') {\n" +
+						"	echo 'Locked'\n" +
+						"}\n" +
+						"echo 'Finish'"
+				));
+
+				FreeStyleProject f = story.j.createFreeStyleProject("f");
+				f.addProperty(new RequiredResourcesProperty("resource1", null, null, null));
+				f.getBuildersList().add(new Shell("sleep 10"));
```

Comment:
You can use a `MockBuilder` (or `TestBuilder`?) which waits for a `Semaphore`. More reliable, portable, and faster.